### PR TITLE
[CBRD-23937] Recreating heap file while TRUNCATE if REUSE_OID	

### PR DESCRIPTION
**Pull request by wrlawodms**
http://jira.cubrid.org/browse/CBRD-23937

See the Jira issue for more information.

As a side note, I've removed "log_does_allow_replication()" condition checking in the code that was commented out. According to my analysis and test, TRUNCATE is logged as one of Statement Based Logging, and it looks working well with destroying heap even if the PK is not dropped. I guess it is a kind of legacy code.

### CODE REVIEW

File: src/object/schema_manager.c
Comment (hornetmj) [inline src/object/schema_manager.c:15890]: In this case, The locator_remove_class_from_index() can be called. Why is it different from before?

File: src/object/schema_manager.c
Comment (hornetmj) [inline src/object/schema_manager.c:15798]: How about moving this line out of for loop? I think it is more readable and It seems to serve the purpose of this for loop better.

File: src/object/schema_manager.c
Comment (hornetmj) [inline src/object/schema_manager.c:15792]: Please explain more detail about this condition and leave it to related Jira issue. Because it is thought to be the most commonly used case. (I remember you shared the reason with me. But, I ask you once again)

**Review by JoohoK — APPROVED** [review]

**Review by hornetmj — APPROVED** [review]

**Comment by hornetmj** [comment]
Plz check and reply to my comment before you merge it. I made the mistake to select 'approve' or 'comment' when I left my comment.

File: src/object/schema_manager.c
Comment (wrlawodms) [reply src/object/schema_manager.c:15798]: Do you mean like the original code?

File: src/object/schema_manager.c
Comment (wrlawodms) [reply src/object/schema_manager.c:15890]: No matter an index is droppable or not, the heap records in a heap with DONT_REUSE_OID have to be deleted record by record. I'll be modified when dealing with the type of heap file.

File: src/object/schema_manager.c
Comment (wrlawodms) [reply src/object/schema_manager.c:15792]: Ok, I've added an overview of these undroppable indexes on the CBRD-23931. When it comes to details, do you mean the description of CBRD-23938 is not enough? To give a precise reason for this need more code analysis.

File: src/storage/heap_file.c
Comment (wrlawodms) [inline src/storage/heap_file.c:6430]: Without this change, when ROLLBACK after TRUNCATE, it generates a dangling overflow file referring to a dropped heap file. This sysop only makes sense when the heap file for a class is never changed and even had an advantage then because it created an overflow heap file only once even with ROLLBACK.

Comment (wrlawodms) [reply src/storage/heap_file.c]: I'll be addressed in another issue because it is a more general bug.

File: src/object/schema_manager.c
Comment (hornetmj) [reply src/object/schema_manager.c:15792]: It's enough. (CBRD-23931)

File: src/object/schema_manager.c
Comment (wrlawodms) [reply src/object/schema_manager.c:15798]: But, if done as you said, it'll still need a condition checking like this to skip a PK referred some FKs. Don't you think it looks redundant?

File: src/storage/system_catalog.c
Comment (hornetmj) [reply src/storage/system_catalog.c:4507]: @wrlawodms I'm in favor of your opinion If it's guaranteed that a record_p always has a hfid.

Comment (wrlawodms) [reply src/storage/system_catalog.c:4507]: `record_p` could be NULL, but then it will be passed in the following `if`. It looks the same logic as the current code.

File: src/object/schema_manager.c
Comment (hornetmj) [reply src/object/schema_manager.c:15798]: I thought that a dummy variable which is the SM_CONSTRAINT_INFO type is used to collect constraints matched in these conditions. It looks a little in-efficient, but more readable. I think the comment just before for loop might be changed or removed if you don't want to change this processing(/* collect index information */).

Comment (wrlawodms) [reply src/object/schema_manager.c:15798]: I see, I'll revise the comment.

**Pull request merged**
